### DISCUSSION TITLE
How can i manage the idiom and notation inside wave components?

### DISCUSSION THREAD

**Discussion by junioraze** [h2oai/wave#2287] [Q&A] [answered]
Hello guys!  
I have an app and i need to change the language of the data picker to pt-br, it is possible? 
![image](https://github.com/h2oai/wave/assets/26543659/c8b43724-b666-4aa0-9688-daa92ead6e9b)

Another point is the 'intl column notation=compact', is there any way to control it and change to 'K' 'M' 'B' (image above in 'mil' 'mi')?
![image](https://github.com/h2oai/wave/assets/26543659/e0626901-ef0c-4341-9554-f81f108a845c)
 
Thanks for read, and sorry my poor english.

**Comment by junioraze** [comment]
*image below*

**Comment by mturoci** [comment]
cc @marek-mihok

**Comment by marek-mihok** [comment]
Hi @junioraze,

>  i need to change the language of the data picker to pt-br, it is possible?

As [discussed here](https://github.com/h2oai/wave/discussions/2106#discussioncomment-6591377), there is no option to do it for components with hardcoded strings. You can create [a new feature request](https://github.com/h2oai/wave/issues/new?assignees=&labels=feature&projects=&template=feature_request.md&title=).

> Another point is the 'intl column notation=compact', is there any way to control it and change to 'K' 'M' 'B' (image above in 'mil' 'mi')?

Could you please provide me with reproducible example? This is as close as I could get:
```py
from h2o_wave import main, app, Q, ui, data


@app('/demo')
async def serve(q: Q):
    q.page['example'] = ui.plot_card(
    box='1 1 4 5',
    title='Interval',
    data=data('year length', 5, rows=[
        ('1991', 1/100),
        ('1995', 1/8),
        ('2002', 1/16),
        ('2003', 1/32),
        ('2010', 1/50),
    ]),
    plot=ui.plot([ui.mark(type='interval', 
                          x='=year', 
                          y_min=0, 
                        y='={{intl length style="unit" unit="mile" type="length" notation="compact" }}'
                          )])
)
    await q.page.save()
```

And what does `'K' 'M' 'B'` mean? Is it a localization (translation) thing or are these just arbitrary units?

**Reply by junioraze** [reply]
Hello @marek-mihok first thanks for reply me and give directions for the first question. About the second one, i think i dont make me clear. When i say "K M B" its about the notation K = thousands, M = million, B=billion.. so if exists a way for me to handle the compact notation in intl.

Code:
```
from h2o_wave import main, app, Q, ui, data

@app("/")
async def serve(q: Q):
    # First time a browser comes to the app
    if not q.client.initialized:
        # Web app metadata.
        q.page["meta"] = ui.meta_card(
            box="",
            # Tab title.
            title="Example #2287",
            # Layout definitions.
            layouts=[
                ui.layout(
                    breakpoint="xs",
                    min_height="100vh",
                    max_width="1200px",
                    zones=[
                        ui.zone("content", size="1"),
                        ui.zone("footer"),
                    ],
                )
            ],
        )


        # Render app body.
        q.page["body"] = ui.plot_card(
            box="content",
            title="Interval",
            data=data(
                "length year",
                5,
                rows=[
                    (10000, "2020-01-01"),
                    (100000, "2021-01-01"),
                    (1000000, "2022-01-01"),
                    (10000000, "2023-01-01"),
                    (100000000, "2024-01-01"),
                ],
            ),
            plot=ui.plot(
                [
                    ui.mark(
                        type="interval",
                        x="=year",
                        y_min=0,
                        y='={{intl length notation="compact" }}',
                        y_scale='log'
                    )
                ]
            ),
        )


        q.client.initialized = True

    await q.page.save()
```

    
![image](https://github.com/h2oai/wave/assets/26543659/f85f1b2e-c6a1-46c6-aee0-28f2a11a9656)

As you can see on y-axis the compact notation handles thousand as "mil" and million as "mi", sometimes in log scales is difficult to understand the difference. Is there any way for me to change this?

**Comment by marek-mihok** [comment, marked as answer]
@junioraze when I run your code, I have the output you expect:

![image](https://github.com/h2oai/wave/assets/23740173/3ed731d7-5d6b-42ea-afab-b472ad3cc039)

The reason why it is showing mi/mil instead of K,M,B is the language of your browser. Our plot implementation uses `navigator.language` prop for localisation of number format which [usually uses language of the browser](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/language).


This is what happens when I override the browser default with `pt-BR` locale:

https://github.com/h2oai/wave/assets/23740173/c4227ae3-e5d1-48ce-bc0f-f5223e053a37

@mturoci wouldn't make sense to prioritise the language of the website by obtaining html `lang` prop [here](https://github.com/h2oai/wave/blob/5fd6aa5447859e9266ed42a6f9b0583281e62c63/ui/src/intl.ts#L21)? Something like this?

```tsx
export const intl = createIntl({
  locale: document.documentElement.getAttribute("lang") || navigator.language,
 ```

**Reply by junioraze** [reply]
Thank y'all, i felt a little confused with the fact that my browser change it. I will try to apply the custom script way, but would be nice if we can configure this unattached to `navigator.language` cuz sometimes we want just to change that exactly behaviour. Thanks again!

**Comment by mturoci** [comment]
> wouldn't make sense to prioritise the language of the website by obtaining html lang prop

Wouldn't that always return `en` since Wave comes with its own index.html?

**Comment by marek-mihok** [comment]
> Wouldn't that always return en since Wave comes with its own index.html?

By default yes, but it can be configured by custom script which would help to resolve this issue:

```py
q.page['meta'].script = ui.inline_script('document.documentElement.setAttribute("lang", "pt-BR");')
```
But this is not ideal solution in my opinion.

What is the meaning of using `navigator.language` for plot's number format in the first place? Browser language does not have to correspond to language of the app.

I think the better idea would be to make Wave index.html `lang` prop configurable and use the language from there.

**Reply by mturoci** [reply]
> it can be configured by custom script

This assumes calling custom script before rendering the plot.

> What is the meaning of using navigator.language for plot's number format in the first place? Browser language does not have to correspond to language of the app.

The reason is internationalization.
